File: Algorithms/p025_Reverse_Nodes_in_k-Group/p025_Reverse_Nodes_in_k-Group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if k<=1:
            return head
        head0 = ListNode(None)
        head0.next = head
        tail = head0
        p1 = head
        while True:
            p2 = p1
            i = 0
            while i<k and p2:
                p2 = p2.next
                i += 1
            if i==k:
                q1 = p1
                q2 = q1.next
                q3 = q2.next
                while True:
                    q2.next = q1
                    if q3==p2:
                        break
                    q1 = q2
                    q2 = q3
                    q3 = q3.next
                tail.next = q2
                tail = p1
                p1 = p2
            else:
                tail.next = p1
                break
        return head0.next

File: Algorithms/p025_Reverse_Nodes_in_k-Group/p025_Reverse_Nodes_in_k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(k<=1) return head;
        ListNode head0(0);
        head0.next = head;
        ListNode* tail = &head0;
        ListNode* p1 = head;
        while(1){
            ListNode* p2 = p1;
            int i = 0;
            for(;i<k&&p2;++i,p2=p2->next) ;
            if(i==k){
                ListNode* q1 = p1;
                ListNode* q2 = q1->next;
                ListNode* q3 = q2->next;
                while(1){
                    q2->next = q1;
                    if(q3==p2) break;
                    q1 = q2;
                    q2 = q3;
                    q3 = q3->next;
                }
                tail->next = q2;
                tail = p1;
                p1 = p2;
            }else{
                tail->next = p1;
                break;
            }
        }
        return head0.next;
    }
};
